src/components/CharacterAssignment.jsx

```diff
@@ -16,85 +16,113 @@ const CharacterAssignment = ({ actors, currentScriptId }) => {
+      let assignedActor = characters[key].actorId;
       if (assignedActor) {
-        newColumns[assignedActor].characterIds.push(fetchedCharacters[key].id);
-        newColumns[assignedActor].lineCount += fetchedCharacters[key].lineCount;
+        newColumns[assignedActor].characterIds.push(characters[key].id);
+        newColumns[assignedActor].lineCount += characters[key].lineCount;
       } else {
-        newColumns.unassignedCharacters.characterIds.push(fetchedCharacters[key].id);
-        newColumns.unassignedCharacters.lineCount += fetchedCharacters[key].lineCount;
+        newColumns.unassignedCharacters.characterIds.push(characters[key].id);
+        newColumns.unassignedCharacters.lineCount += characters[key].lineCount;
       }
     }
-    setColumns({ ...newColumns });
-    setCharacters(fetchedCharacters);
+    return newColumns;
   };
 
   /**
-   * Create the actor and unassigned columns, or update them when the number of actors changes
-   * @param actors - Array of objects with properties of name and id
+   * Creates the actor columns, all empty at this stage
+   * @param {array} actors Array of objects with properties of name and id
+   * @returns {object} newColumns: Object of actor column objects, actorIds: array of actor ids, represents column order
    */
-    // Create and update the columns for each of the actors and an Unassigned Characters column
-  const createOrUpdateActorColumns = (actors) => {
-      if (!currentScriptId && actors.length == 0) {
-        // Initial setup of actor columns, no script is selected
-        createInitialActorColumns(actors);
-        return;
+  const createInitialActorColumns = (actors) => {
+    const newColumns = {
+      'unassignedCharacters': {
+        id: 'unassignedCharacters',
+        title: 'Unassigned Characters',
+        characterIds: [],
+        lineCount: 0
       }
-
-      // Script is already set, only add or remove actors
-      const newColumns = { ...columns };
-      const newColumnList = [...columnList];
+    };
+    const actorIds = [];
+    actors.forEach((actor) => {
+      newColumns[actor.id] = {
+        id: actor.id.toString(),
+        title: actor.name,
+        characterIds: [],
+        lineCount: 0
+      };
+      actorIds.push(actor.id.toString());
+    });
+    return {newColumns, actorIds}
+  };
+  /**
+   * Update the actor columns when the number of actors changes and sets the new column values and order
+   * @param actors Array of objects with properties of name and id
+   */
+  const updateActorColumns = (actors) => {
+      const oldColumns = { ...columns };
+      const oldColumnOrder = [...columnOrder];
       // If there are more actors than columns, add a column
-      if (actors.length > newColumnList.length) {
-        addActorColumn(actors, newColumns, newColumnList);
+      if (actors.length > oldColumnOrder.length) {
+        const { newColumns, newColumnOrder} = addActorColumn(actors, oldColumns, oldColumnOrder);
```

Comment:
nit: Having three copies of each variable, (`columns`, `oldColumns`, and `newColumns`) makes it a little harder to keep track of.  

One way to reduce the duplication could be to inline the object/array copying:
```js
const { newColumns, newColumnOrder} = addActorColumn(actors, { ...columns }, [...columnOrder]);
```
or `addActorColumn()` could the copying internally such that it returns new object/arrays.

Reply:
It would not have occurred to me to do that inline and that is an excellent solution. Thank you!